core/site.go

```diff
@@ -387,6 +423,143 @@ func (site *Site) sitePower(totalChargePower float64) (float64, error) {
+				site.log.TRACE.Printf("lm: loadpoint %s in StateC gets %.1fA assigned. Remaining budget: %.1fA", curLp.Name(), curLp.lmCurrentAllowance, lmCurrentBudget)
+			} else {
+				// not enough budget for at least minCurrent. Check if there are more LP's left. If so, take their current
+				// which will effectively disable the rest of LP's
+				if lpStateBcnt+lpStateCcnt > 1 {
+					// use remaining power
+					lpCurrent := math.Min(lmCurrentBudget, curLp.effectiveCurrent()*chargingOffset)
+					curLp.setLMParams(true, lpCurrent)
+					lmCurrentBudget -= lpCurrent
+					site.log.TRACE.Printf("lm: loadpoint %s in StateC gets %.1fA assigned (at cost of remaining LP's). Remaining budget: %.1fA", curLp.Name(), curLp.lmCurrentAllowance, lmCurrentBudget)
+				} else {
+					// nothing left ... set to 0
+					curLp.setLMParams(true, 0.0)
+					site.log.TRACE.Printf("lm: loadpoint %s in StateB gets no current assigned, too less budget remaining. Remaining budget: %.1fA", curLp.Name(), lmCurrentBudget)
+				}
+			}
+			lpStateCcnt--
+		}
+	}
+	// there might be budget of current left after the distribution.
+	// if so, assign the remaining budget to stateC lp which have lower assignment than MaxCurrent
+	site.log.DEBUG.Printf("lm: first loop, remaining budget: %.1fA\n", lmCurrentBudget)
+	if lmCurrentBudget > 0.0 {
+		for _, curLp := range site.loadpoints {
+			if curLp.GetStatus() == api.StatusC && curLp.lmCurrentAllowance < curLp.GetMaxCurrent() {
+				offset := math.Min(curLp.GetMaxCurrent()-curLp.lmCurrentAllowance, lmCurrentBudget)
+				curLp.setLMParams(true, curLp.lmCurrentAllowance+offset)
+				site.log.TRACE.Printf("lm: add %.1fA to lp from remaining budget %.1fA to lp %s\n", offset, lmCurrentBudget, curLp.Name())
+				lmCurrentBudget -= offset
+				if lmCurrentBudget == 0.0 {
+					break
+				}
+			}
+		}
+	}
+	// finally, if current is left, assign to StatusB LP
+	if lmCurrentBudget > 0.0 {
+		for _, curLp := range site.loadpoints {
+			if curLp.GetStatus() == api.StatusB && curLp.lmCurrentAllowance < curLp.GetMaxCurrent() {
+				offset := math.Min(curLp.GetMaxCurrent()-curLp.lmCurrentAllowance, lmCurrentBudget)
+				curLp.setLMParams(true, curLp.lmCurrentAllowance+offset)
+				site.log.TRACE.Printf("lm: add %.1fA to lp from remaining budget %.1fA to lp %s\n", offset, lmCurrentBudget, curLp.Name())
+				lmCurrentBudget -= offset
+				if lmCurrentBudget == 0.0 {
+					break
+				}
+			}
+		}
+	}
+
+	// finally ensure the limitation by setting the limit using their effectiveCurrent()
+	// if lmCurrentAllowance is lower than actual current
+	if enforce {
+		for _, curLp := range site.loadpoints {
+			if (curLp.GetStatus() == api.StatusC || curLp.GetStatus() == api.StatusB) &&
+				(curLp.lmCurrentAllowance < curLp.effectiveCurrent()) {
+				if err := curLp.setLimit(curLp.effectiveCurrent(), false); err != nil { // if lm hits, it will set the force flag and apply the lm current
+					site.log.ERROR.Println(err)
+				}
+			}
+		}
+	}
+}
```

Comment:
Puh. Horrorfunktion. Das müssen wir irgendwie abschichten.

Reply:
Ich versuche zumindest viele Kommentare unterzubringen :-). 
Kann gerne umgebaut werden, ist aktuell aber eine Sequenz aus
1) ermitteln benötigten Stroms
2) Verteilen auf die LP nach Anforderung
3) Verteilen des Rests vom Budget auf die LP mit mehr Bedarf. (Entsteht, wenn hintere LP in der Liste weniger Strom anfordern als sie bei Gleichverteilung erhalten würden)

Reply:
Man kann das auch in ein Modul auslagern, welches das Lastmanagement auf die LP anwendet. Wenn man das tut, hätte man auch die Chance verschiedere Strategien zu verwenden (LM Interface) - im Hinblick auf Loadbalancing.

Reply:
Schau Dir mal meinen Vorschlag an. Da liegt die Logik weiter in den Ladepunkten, sie müssen ihre Bedarfe aber beim Circuit anmelden. Der entscheidet was erlaubt ist. Eine Verteillogik hat eigentlich mit LM erstmal gar nichts zu tun sondern ist nochmal eine Funktion on Top die man zusätzlich kapseln kann (und muss um später z.B. auch SoC-bezogene Parameter mit einzubauen).

Du hast Dir wirklich viel Mühe gemacht! Ich glaube aber, dass hier noch zu viel ineinander verschachtelt und verwoben ist und die Strukturen auch im Code auseinander gezogen werden sollten.

Reply:
@andig besteht die Möglichkeit das zu besprechen? Dann liesse sich ggf besser abstimmen was gemeint ist, und was der beste Weg ist.

Reply:
Gerne- magst Du Dich mal bei Slack melden?

Reply:
Gerne. Der Link von der HOmepage für den Space ist aber abgelaufen. 
Gerne auch per PM senden.

Reply:
> Gerne auch per PM senden.

Ähhh- aber wie?

Reply:
Schau mal https://evcc.io/slack